### PR TITLE
EN-4321 : add in termui warning when node needs to be updated

### PR DESCRIPTION
**Pull request by miiu96**
Added a new feature in termui that checks if a new software stable tag is available

### CODE REVIEW

File: cmd/node/main.go [hunk 1]
```diff
@@ -701,6 +702,11 @@ func startNode(ctx *cli.Context, log *logger.Logger, version string) error {
 		return err
 	}
 
+	softwareVersionChecker, err := softwareVersion.NewSoftwareVersionChecker(coreComponents.StatusHandler)
```
Comment (sasurobert) [inline cmd/node/main.go]: can't we have a factory for all these metrics ?

File: core/statistics/softwareVersion/softwareVersion.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			case <-time.After(checkInterval):
+				svc.readLatestStableVersion()
+			}
+		}
+	}()
+}
+
+func (svc *softwareVersionChecker) readLatestStableVersion() {
+	tagVersion, err := readJSONFromUrl(stableTagLocation)
+	if err == nil {
```
Comment (sasurobert) [inline core/statistics/softwareVersion/softwareVersion.go]: if err != nil log Error and return. After that call the setStringValue

File: statusHandler/view/termuic/termuiRenders/widgetsRender.go [hunk 3]
```diff
@@ -106,9 +106,11 @@ func (wr *WidgetsRender) prepareInstanceInfo() {
 	rows := make([][]string, numRows)
 
 	appVersion := wr.presenter.GetAppVersion()
+	needUpdate, latestStableVersion := wr.presenter.CheckSoftwareVersion()
 	rows[0] = []string{fmt.Sprintf("App version: %s", appVersion)}
-	if strings.Contains(appVersion, core.UnVersionedAppString) {
-		wr.instanceInfo.RowStyles[0] = ui.NewStyle(ui.ColorRed)
+	if needUpdate {
```
Comment (sasurobert) [inline statusHandler/view/termuic/termuiRenders/widgetsRender.go]: add empty line

File: core/statistics/softwareVersion/softwareVersion.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+type softwareVersionChecker struct {
+	statusHandler core.AppStatusHandler
+}
+
+// NewSoftwareVersionChecker will create a object for software  version checker
+func NewSoftwareVersionChecker(appStatusHandler core.AppStatusHandler) (*softwareVersionChecker, error) {
+	if appStatusHandler == nil || appStatusHandler.IsInterfaceNil() {
+		return nil, appStatusPolling.ErrNilAppStatusHandler
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion.go]: declare the error in the same package. No need to import another package just for this

File: core/statistics/softwareVersion/softwareVersion.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+func readJSONFromUrl(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(resp.Body)
+	respByte := buf.Bytes()
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion.go]: respBytes*

File: core/statistics/softwareVersion/softwareVersion.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+	var tag tagVersion
+	if err := json.Unmarshal(respByte, &tag); err != nil {
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion.go]: no need for `:` as `err` is not a new variable

File: core/statistics/softwareVersion/softwareVersion.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	}
+}
+
+func readJSONFromUrl(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion.go]: refactor this to :
```GOLANG
defer func() {
		err := resp.Body.Close()
		if err != nil {
			log(....)
		}
	}()
````
as the error which can be returned by the `Close()` method is not treated (GoLand warning)

File: core/statistics/softwareVersion/softwareVersion_test.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewSoftwareVersionChecker_NilStatusHandlerShouldErr(t *testing.T) {
+	t.Parallel()
+
+	softwareChecker, err := NewSoftwareVersionChecker(nil)
+
+	assert.Nil(t, softwareChecker)
+	assert.Equal(t, appStatusPolling.ErrNilAppStatusHandler, err)
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion_test.go]: also expect here the error you will declare in the same package

File: core/statistics/softwareVersion/softwareVersion_test.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package softwareVersion
+
+import (
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/consensus/mock"
```
Comment (bogdan-rosianu) [inline core/statistics/softwareVersion/softwareVersion_test.go]: why you use the AppStatusHandlerMock from consensus package ? there is a stub for this component in the core package

File: cmd/node/main.go (hunk 1 above)
Comment (bogdan-rosianu) [reply cmd/node/main.go]: +1

File: core/statistics/softwareVersion/softwareVersion.go (hunk 2 above)
Comment (miiu96) [reply core/statistics/softwareVersion/softwareVersion.go]: done.

File: statusHandler/view/termuic/termuiRenders/widgetsRender.go (hunk 3 above)
Comment (miiu96) [reply statusHandler/view/termuic/termuiRenders/widgetsRender.go]: done.

File: core/statistics/softwareVersion/softwareVersion.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+type tagVersion struct {
+	TagVersion string `json:"tag_name"`
+}
+
+type softwareVersionChecker struct {
+	statusHandler core.AppStatusHandler
+}
+
+// NewSoftwareVersionChecker will create a object for software  version checker
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/statistics/softwareVersion/softwareVersion.go]: done.

File: core/statistics/softwareVersion/softwareVersion.go (hunk 7 above)
Comment (miiu96) [reply core/statistics/softwareVersion/softwareVersion.go]: done.

File: core/statistics/softwareVersion/softwareVersion_test.go (hunk 8 above)
Comment (miiu96) [reply core/statistics/softwareVersion/softwareVersion_test.go]: done.

File: core/statistics/softwareVersion/softwareVersion_test.go (hunk 9 above)
Comment (miiu96) [reply core/statistics/softwareVersion/softwareVersion_test.go]: done.

File: cmd/node/main.go (hunk 1 above)
Comment (miiu96) [reply cmd/node/main.go]: done.

**Pull request merged**